File: java/minimumAddToMakeParenthesesValid.java
class Solution {
    public int minAddToMakeValid(String S) {
        int numNeeded = 0; //number of parentheses needed to insert to make string valid; updates at the end of every 'segment'
        int openParens = 0; //number of open parentheses in the current 'segment'
        int closeParens = 0;
        //traverse the string and end a 'segment' when the number of close parentheses is greater than the number of open parentheses
        //increment numNeeded at the end of each such segment by one, which corresponds to the open parenthesis needed to make the segment valid
        for (int i = 0; i < S.length(); i++) {
            char c = S.charAt(i);
            if (c == '(') {
                openParens++;
            } else {
                closeParens++;
            }
            if (closeParens > openParens) {
                numNeeded++;
                openParens = 0;
                closeParens = 0;
            }
        }
        //add to numNeeded the number of close parentheses necessary to close out the remaining excess open parentheses
        numNeeded += openParens - closeParens;
        return numNeeded;
    }
}

File: python/minimum_add_to_make_parentheses_valid.py
class Solution:
    def minAddToMakeValid(self, S: str) -> int:
        num_needed = 0 #keeps track of number of parentheses necessary to make the string valid; added to for each string segment
        num_open_parens = 0 #keeps track of number of open parentheses in the current segment
        num_close_parens = 0
        for i, ch in enumerate(S):
            if ch == "(": num_open_parens += 1
            else: num_close_parens += 1
            #if there are more close parentheses than open parentheses, end the current segment, add the difference to num_needed, and refresh the two counters
            if ch == ")" and num_close_parens > num_open_parens: 
                num_needed += num_close_parens - num_open_parens
                num_open_parens, num_close_parens = 0, 0
        num_needed += num_open_parens - num_close_parens
        return num_needed